<a href="https://colab.research.google.com/github/Abiads/python-10-days-XebAppi/blob/main/DAY4_PYTHON.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

DAILY REPORT  
DATE: 24 March 2026  

WORK DONE  
• Studied **CLASS & OBJECT** concepts to structure data and build reusable code components  
• Implemented **__init__ METHOD** to initialize object data and manage instance variables  
• Used **INHERITANCE** to reuse parent class properties and reduce code duplication  
• Applied **ENCAPSULATION** to protect data using controlled access methods  
• Practiced **POLYMORPHISM** to handle multiple behaviors using same function interface  
• Explored **SPECIAL METHODS** like __str__ for better object representation  
• Installed **PACKAGES** using pip to extend program functionality with external libraries  
• Used **LIBRARIES** like os and datetime to handle system operations and time data  
• Implemented **RE MODULE** for pattern matching and input validation  
• Organized code using **PACKAGES** to improve modular structure and maintainability  

KEY LEARNINGS & OBSERVATIONS  
• **OOP CONCEPTS** help in writing clean and scalable code for real projects  
• **__init__ METHOD** is important for automatic object setup during creation  
• **INHERITANCE** reduces repetition which saves time in large applications  
• **ENCAPSULATION** improves DATA SECURITY and prevents unwanted changes  
• **POLYMORPHISM** makes code flexible and easy to extend in future  
• **SPECIAL METHODS** improve readability when working with objects  
• **LIBRARIES** save development time by providing ready-made solutions  
• **pip INSTALLATION** is essential to use external tools and frameworks  
• **REGEX (re module)** is powerful for VALIDATION and text processing  
• **PACKAGES** help manage large projects by splitting code into modules  

In [1]:
from abc import ABC, abstractmethod
import math

# 2. Define an abstract base class named Shape that inherits from ABC.
class Shape(ABC):
    # 3. Within the Shape class, define two abstract methods: area() and perimeter().
    @abstractmethod
    def area(self):
        pass

    @abstractmethod
    def perimeter(self):
        pass

# 4. Define a concrete class named Circle that inherits from Shape.
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    # 5. Implement the area() and perimeter() methods within the Circle class.
    def area(self):
        return math.pi * self.radius ** 2

    def perimeter(self):
        return 2 * math.pi * self.radius

# 6. Create an instance of the Circle class, for example, with a radius of 5.
circle_instance = Circle(radius=5)

# 7. Call the area() and perimeter() methods on the Circle instance and print their results.
print(f"Circle Area: {circle_instance.area():.2f}")
print(f"Circle Perimeter: {circle_instance.perimeter():.2f}")

Circle Area: 78.54
Circle Perimeter: 31.42


In [2]:
import math

# 1. Define a class named Engine with an __init__ method that takes horsepower and fuel_type as arguments and stores them as attributes. Add a method named start() that prints a message indicating the engine is starting.
class Engine:
    def __init__(self, horsepower, fuel_type):
        self.horsepower = horsepower
        self.fuel_type = fuel_type

    def start(self):
        print(f"The {self.fuel_type} engine with {self.horsepower} HP is starting.")

# 2. Define a class named Car with an __init__ method that takes make, model, and engine_type (e.g., 'V6', 'Electric') as arguments. Inside the Car's __init__, create an instance of the Engine class based on the engine_type (e.g., if 'V6', instantiate Engine(horsepower=300, fuel_type='Gasoline')). Store this Engine instance as an attribute of the Car (e.g., self.engine).
class Car:
    def __init__(self, make, model, engine_type):
        self.make = make
        self.model = model
        if engine_type == 'V6':
            self.engine = Engine(horsepower=300, fuel_type='Gasoline')
        elif engine_type == 'Electric':
            self.engine = Engine(horsepower=250, fuel_type='Electric')
        else:
            self.engine = Engine(horsepower=150, fuel_type='Unknown') # Default or custom engine

    # 3. Add a method to the Car class named drive() that prints a message indicating the car is driving and then calls the start() method of its engine attribute.
    def drive(self):
        print(f"The {self.make} {self.model} is driving.")
        self.engine.start()

# 4. Create an instance of the Car class, for example, a 'Toyota Camry' with a 'V6' engine.
my_car = Car(make='Toyota', model='Camry', engine_type='V6')

# 5. Call the drive() method on the Car instance to demonstrate the composition.
my_car.drive()

electric_car = Car(make='Tesla', model='Model 3', engine_type='Electric')
electric_car.drive()

The Toyota Camry is driving.
The Gasoline engine with 300 HP is starting.
The Tesla Model 3 is driving.
The Electric engine with 250 HP is starting.


In [3]:
import math

# 1. Define a base class named Animal with an __init__ method that takes a name and stores it as an attribute. Add a method named speak() that prints a generic message like 'The animal makes a sound.'.
class Animal:
    def __init__(self, name):
        self.name = name

    def speak(self):
        print(f"The {self.name} makes a sound.")

# 2. Define a derived class named Dog that inherits from Animal. The Dog class's __init__ method should call the parent class's __init__ using super().__init__(name) and also add a breed attribute.
class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name)
        self.breed = breed

    # 3. Override the speak() method in the Dog class to print a specific message for a dog, such as 'Woof! Woof!'.
    def speak(self):
        print(f"{self.name}, the {self.breed} says: Woof! Woof!")

# 4. Define another derived class named Cat that also inherits from Animal. The Cat class's __init__ method should similarly call the parent's __init__ and add a color attribute.
class Cat(Animal):
    def __init__(self, name, color):
        super().__init__(name)
        self.color = color

    # 5. Override the speak() method in the Cat class to print a specific message for a cat, such as 'Meow.'.
    def speak(self):
        print(f"{self.name}, the {self.color} cat, says: Meow.")

# 6. Create an instance of the Animal class, a Dog class, and a Cat class.
animal_instance = Animal(name="Generic Animal")
dog_instance = Dog(name="Buddy", breed="Golden Retriever")
cat_instance = Cat(name="Whiskers", color="Orange")

# 7. Call the speak() method on each instance to demonstrate method overriding.
print("Demonstrating Method Overriding:")
animal_instance.speak()
dog_instance.speak()
cat_instance.speak()


Demonstrating Method Overriding:
The Generic Animal makes a sound.
Buddy, the Golden Retriever says: Woof! Woof!
Whiskers, the Orange cat, says: Meow.


In [4]:
import math

# 1. Define a base class named Flyable with a method fly() that prints 'Flying high!'.
class Flyable:
    def fly(self):
        print('Flying high!')

# 2. Define another base class named Swimmable with a method swim() that prints 'Swimming gracefully!'.
class Swimmable:
    def swim(self):
        print('Swimming gracefully!')

# 3. Define a derived class named Duck that inherits from both Flyable and Swimmable.
class Duck(Flyable, Swimmable):
    def __init__(self, name):
        self.name = name
        print(f"Hello, I am {self.name}, a duck.")

# 4. Create an instance of the Duck class.
duck_instance = Duck("Daffy")

# 5. Call the fly() method on the Duck instance.
duck_instance.fly()

# 6. Call the swim() method on the Duck instance.
duck_instance.swim()

# 7. Print the Method Resolution Order (MRO) for the Duck class using Duck.mro() to show the order in which Python searches for methods in multiple inheritance.
print("\nMethod Resolution Order (MRO) for Duck:")
print(Duck.mro())

Hello, I am Daffy, a duck.
Flying high!
Swimming gracefully!

Method Resolution Order (MRO) for Duck:
[<class '__main__.Duck'>, <class '__main__.Flyable'>, <class '__main__.Swimmable'>, <class 'object'>]


## Summary:

### Q&A
The key OOP concepts covered include Abstraction, Composition, Method Overriding, and Multiple Inheritance.
*   **Abstraction** is useful for defining a common interface or blueprint for a set of classes, ensuring that all subclasses implement certain methods. It helps in managing complexity by hiding implementation details and showing only essential features.
*   **Composition** is useful for building complex objects from simpler ones, representing a "has-a" relationship. It promotes code reuse and flexibility by allowing independent development and testing of component parts.
*   **Method Overriding** is useful when a subclass needs to provide a specific implementation for a method that is already defined in its parent class. This allows for specialized behavior while maintaining a common interface defined by the parent.
*   **Multiple Inheritance** (though generally used cautiously due to complexity) is useful when a class needs to inherit functionalities from multiple distinct parent classes, combining their behaviors into a single entity.

### Data Analysis Key Findings
*   **Abstraction** was successfully demonstrated using Python's `abc` module. An abstract base class `Shape` was defined with abstract methods `area()` and `perimeter()`, which were then implemented by a concrete subclass `Circle`. For a `Circle` with radius 5, the calculated area was 78.54 and the perimeter was 31.42.
*   **Composition** was illustrated by creating a `Car` class that "has an" `Engine` object. The `Car`'s `drive()` method internally called the `Engine`'s `start()` method, showing how one object utilizes functionality from its contained objects. For example, a 'Toyota Camry' with a 'V6' engine started a 'Gasoline engine with 300 HP'.
*   **Method Overriding** was shown with an `Animal` base class and derived classes `Dog` and `Cat`. Each derived class provided its own specific implementation for the `speak()` method, overriding the generic `Animal` version. For instance, a `Dog` named "Buddy" barked "Woof! Woof!", while a `Cat` named "Whiskers" meowed "Meow."
*   **Multiple Inheritance** was demonstrated by a `Duck` class inheriting methods (`fly()` and `swim()`) from two separate base classes, `Flyable` and `Swimmable`. The Method Resolution Order (MRO) for `Duck` was confirmed as `[<class '__main__.Duck'>, <class '__main__.Flyable'>, <class '__main__.Swimmable'>, <class 'object'>]`, indicating the order Python searches for methods.



In [5]:
import os

# 1. Create a main directory for your custom package
project_root = 'my_custom_package'
if not os.path.exists(project_root):
    os.makedirs(project_root)
    print(f"Created directory: {project_root}")
else:
    print(f"Directory already exists: {project_root}")

# 2. Inside the my_custom_package directory, create another subdirectory that will serve as the actual Python package.
package_dir = os.path.join(project_root, 'my_package')
if not os.path.exists(package_dir):
    os.makedirs(package_dir)
    print(f"Created directory: {package_dir}")
else:
    print(f"Directory already exists: {package_dir}")

# 3. Inside the my_package directory, create an empty file named __init__.py.
init_file_path = os.path.join(package_dir, '__init__.py')
if not os.path.exists(init_file_path):
    with open(init_file_path, 'w') as f:
        pass # Create an empty file
    print(f"Created empty file: {init_file_path}")
else:
    print(f"File already exists: {init_file_path}")

Created directory: my_custom_package
Created directory: my_custom_package/my_package
Created empty file: my_custom_package/my_package/__init__.py


In [6]:
import os

# Define the path for greetings.py
greetings_file_path = os.path.join(package_dir, 'greetings.py')

# Content for greetings.py
greetings_content = """def say_hello(name):
    return f"Hello, {name}! Welcome to my custom package."

def say_goodbye(name):
    return f"Goodbye, {name}! Hope to see you again soon."
"""

# Write the content to greetings.py
if not os.path.exists(greetings_file_path):
    with open(greetings_file_path, 'w') as f:
        f.write(greetings_content)
    print(f"Created file: {greetings_file_path} with functions.")
else:
    print(f"File already exists: {greetings_file_path}. Content will be updated if different.")
    with open(greetings_file_path, 'w') as f:
        f.write(greetings_content)
    print(f"Updated file: {greetings_file_path} with functions.")

Created file: my_custom_package/my_package/greetings.py with functions.


In [7]:
class Father:
    def __init__(self, name):
        self.name = name
        self.hobby_name = "fishing"
        print(f"Father {self.name} is initialized.")

    def hobby(self):
        print(f"{self.name} enjoys {self.hobby_name}.")

class Mother:
    def __init__(self, name):
        self.name = name
        self.skill_name = "cooking"
        print(f"Mother {self.name} is initialized.")

    def skill(self):
        print(f"{self.name} is skilled at {self.skill_name}.")

class Child(Father, Mother):
    def __init__(self, name, father_name, mother_name):
        self.name = name # Child's own name
        # Initialize parent classes explicitly for clarity in multiple inheritance
        Father.__init__(self, father_name)
        Mother.__init__(self, mother_name)
        print(f"Child {self.name} is initialized, with father {father_name} and mother {mother_name}.")

    def play(self):
        print(f"Child {self.name} is playing.")

# Create an instance of the Child class
child_instance = Child(name="Alice", father_name="Bob", mother_name="Carol")

# Call the hobby() method (from Father) and the skill() method (from Mother) on the Child instance
child_instance.hobby()
child_instance.skill()

# Call the play() method on the Child instance
child_instance.play()

# Print the Method Resolution Order (MRO) for the Child class
print("\nMethod Resolution Order (MRO) for Child:")
print(Child.mro())

Father Bob is initialized.
Mother Carol is initialized.
Child Carol is initialized, with father Bob and mother Carol.
Carol enjoys fishing.
Carol is skilled at cooking.
Child Carol is playing.

Method Resolution Order (MRO) for Child:
[<class '__main__.Child'>, <class '__main__.Father'>, <class '__main__.Mother'>, <class 'object'>]


In [8]:
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age
        print(f"Person {self.name} is initialized with age {self.age}.")

    def get_person_details(self):
        print(f"Name: {self.name}, Age: {self.age}")

class Employee(Person):
    def __init__(self, name, age, employee_id):
        super().__init__(name, age)  # Call the parent class's constructor
        self.employee_id = employee_id
        print(f"Employee {self.name} (ID: {self.employee_id}) is initialized.")

    def get_employee_details(self):
        print(f"Name: {self.name}, Age: {self.age}, Employee ID: {self.employee_id}")

# Create an instance of the Employee class
employee_instance = Employee(name='Alice', age=30, employee_id='EMP001')

# Call the get_employee_details() method on the Employee instance
print("\nDemonstrating Employee details:")
employee_instance.get_employee_details()

Person Alice is initialized with age 30.
Employee Alice (ID: EMP001) is initialized.

Demonstrating Employee details:
Name: Alice, Age: 30, Employee ID: EMP001


In [9]:
class Shape:
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        print('Calculating area of a shape.')
        return 0

class Rectangle(Shape):
    def __init__(self, width, height):
        super().__init__(width, height)

    def area(self):
        return self.width * self.height

# Create an instance of the Shape class
shape_instance = Shape(10, 5)

# Create an instance of the Rectangle class
rectangle_instance = Rectangle(10, 5)

# Call the area() method on both instances and print their results
print(f"Shape Area: {shape_instance.area()}")
print(f"Rectangle Area: {rectangle_instance.area()}")

Calculating area of a shape.
Shape Area: 0
Rectangle Area: 50


In [10]:
class Company:
    # 2. Define a class variable named industry
    industry = 'Technology'

    # 3. Implement the __init__ method for the Company class
    def __init__(self, name, founded_year):
        self.name = name
        self.founded_year = founded_year

# 4. Create two instances of the Company class
company1 = Company('Tech Innovators Inc.', 2005)
company2 = Company('Future Systems Ltd.', 2010)

# 5. Print the industry (class variable), name, and founded_year for both
print("\n--- Initial State ---")
print(f"Company 1: Name = {company1.name}, Founded = {company1.founded_year}, Industry = {company1.industry}")
print(f"Company 2: Name = {company2.name}, Founded = {company2.founded_year}, Industry = {company2.industry}")

# 6. Modify the industry class variable directly through the Company class
Company.industry = 'Software Development'

# 7. Print the industry variable again for both company1 and company2
print("\n--- After modifying class variable via Company class ---")
print(f"Company 1: Name = {company1.name}, Founded = {company1.founded_year}, Industry = {company1.industry}")
print(f"Company 2: Name = {company2.name}, Founded = {company2.founded_year}, Industry = {company2.industry}")

# 8. Attempt to modify the industry variable through company1
company1.industry = 'Hardware Manufacturing'

# 9. Print the industry variable for company1 and company2 one more time
print("\n--- After modifying instance variable via company1 ---")
print(f"Company 1: Name = {company1.name}, Founded = {company1.founded_year}, Industry = {company1.industry}")
print(f"Company 2: Name = {company2.name}, Founded = {company2.founded_year}, Industry = {company2.industry}")



--- Initial State ---
Company 1: Name = Tech Innovators Inc., Founded = 2005, Industry = Technology
Company 2: Name = Future Systems Ltd., Founded = 2010, Industry = Technology

--- After modifying class variable via Company class ---
Company 1: Name = Tech Innovators Inc., Founded = 2005, Industry = Software Development
Company 2: Name = Future Systems Ltd., Founded = 2010, Industry = Software Development

--- After modifying instance variable via company1 ---
Company 1: Name = Tech Innovators Inc., Founded = 2005, Industry = Hardware Manufacturing
Company 2: Name = Future Systems Ltd., Founded = 2010, Industry = Software Development


In [11]:
class MathUtils:
    # 2. Define a static method named add
    @staticmethod
    def add(a, b):
        # 4. The add method should return the sum of a and b
        return a + b

# 5. Call the add static method using the class name and print the result
result = MathUtils.add(5, 3)
print(f"The sum of 5 and 3 using MathUtils.add is: {result}")

result2 = MathUtils.add(10, 20)
print(f"The sum of 10 and 20 using MathUtils.add is: {result2}")

The sum of 5 and 3 using MathUtils.add is: 8
The sum of 10 and 20 using MathUtils.add is: 30


In [16]:
class College:
    # 2. Define a class variable named total_colleges
    total_colleges = 0

    # 3. Implement the __init__ method for the College class
    def __init__(self, name, city):
        self.name = name
        self.city = city
        # Increment the total_colleges class variable
        College.total_colleges += 1
        print(f"College instance created: {self.name} in {self.city}. Total colleges: {College.total_colleges}")

    # 4. Define a class method named create_from_string
    @classmethod
    def create_from_string(cls, college_info_string):
        # 6. Parse the college_info_string
        name, city = college_info_string.split('-')
        # Use cls(name, city) to create and return a new College instance
        return cls(name, city)

# 7. Create two instances of the College class using the standard constructor
print("\n--- Creating instances with standard constructor ---")
college1 = College('SRIT', 'MNIT')
college2 = College('IIT', 'UIT')

# 8. Create a third instance using the create_from_string class method
print("\n--- Creating instance with class method ---")
college3 = College.create_from_string('Jabalpur-MP')

# 9. Print the name and city for each instance
print("\n--- College Details ---")
print(f"College 1: {college1.name}, City: {college1.city}")
print(f"College 2: {college2.name}, City: {college2.city}")
print(f"College 3: {college3.name}, City: {college3.city}")

# 10. Print the total_colleges class variable
print("\n--- Total Colleges ---")
print(f"Total colleges created (via class): {College.total_colleges}")
print(f"Total colleges created (via instance college1): {college1.total_colleges}")
print(f"Total colleges created (via instance college3): {college3.total_colleges}")


--- Creating instances with standard constructor ---
College instance created: SRIT in MNIT. Total colleges: 1
College instance created: IIT in UIT. Total colleges: 2

--- Creating instance with class method ---
College instance created: Jabalpur in MP. Total colleges: 3

--- College Details ---
College 1: SRIT, City: MNIT
College 2: IIT, City: UIT
College 3: Jabalpur, City: MP

--- Total Colleges ---
Total colleges created (via class): 3
Total colleges created (via instance college1): 3
Total colleges created (via instance college3): 3


In [17]:
import shutil
import os

# 2. Create a dummy file named source_file.txt with some content in the current working directory.
source_file_name = 'source_file.txt'
with open(source_file_name, 'w') as f:
    f.write('This is a dummy file for shutil demonstration.')
print(f"Created: {source_file_name}")

# 3. Create a directory named destination_copy_dir.
dest_copy_dir = 'destination_copy_dir'
os.makedirs(dest_copy_dir, exist_ok=True)
print(f"Created directory: {dest_copy_dir}")

# 4. Use shutil.copy() to copy source_file.txt into destination_copy_dir. Print a confirmation message.
shutil.copy(source_file_name, dest_copy_dir)
print(f"Copied '{source_file_name}' to '{dest_copy_dir}'")

# 5. Create a directory named destination_move_dir.
dest_move_dir = 'destination_move_dir'
os.makedirs(dest_move_dir, exist_ok=True)
print(f"Created directory: {dest_move_dir}")

# 6. Use shutil.move() to move source_file.txt (from its original location, or its current location if moved in previous steps) into destination_move_dir.
# Note: After the copy operation, source_file.txt still exists in the current directory.
shutil.move(source_file_name, dest_move_dir)
print(f"Moved '{source_file_name}' to '{dest_move_dir}'")

# 7. Create another dummy directory named dir_to_remove and create a file inside it (e.g., dir_to_remove/test.txt) to demonstrate rmtree on a non-empty directory.
dir_to_remove = 'dir_to_remove'
os.makedirs(dir_to_remove, exist_ok=True)
with open(os.path.join(dir_to_remove, 'test.txt'), 'w') as f:
    f.write('This file is inside a directory to be removed.')
print(f"Created directory '{dir_to_remove}' with 'test.txt' inside.")

# 8. Use shutil.rmtree() to remove dir_to_remove. Print a confirmation message.
shutil.rmtree(dir_to_remove)
print(f"Removed directory '{dir_to_remove}' using shutil.rmtree()")

# 9. Clean up any remaining created files or directories using os.remove() or os.rmdir() if necessary (e.g., destination_copy_dir and destination_move_dir).
# Ensure to check if they exist before attempting to remove them.
# The original source_file.txt has been moved.

# Clean up destination_copy_dir
if os.path.exists(dest_copy_dir):
    # The copied file is inside, so rmtree is appropriate for non-empty directory
    shutil.rmtree(dest_copy_dir)
    print(f"Cleaned up directory: {dest_copy_dir}")

# Clean up destination_move_dir
if os.path.exists(dest_move_dir):
    # The moved file is inside, so rmtree is appropriate for non-empty directory
    shutil.rmtree(dest_move_dir)
    print(f"Cleaned up directory: {dest_move_dir}")

print("Shutil demonstration complete and cleanup performed.")

Created: source_file.txt
Created directory: destination_copy_dir
Copied 'source_file.txt' to 'destination_copy_dir'
Created directory: destination_move_dir
Moved 'source_file.txt' to 'destination_move_dir'
Created directory 'dir_to_remove' with 'test.txt' inside.
Removed directory 'dir_to_remove' using shutil.rmtree()
Cleaned up directory: destination_copy_dir
Cleaned up directory: destination_move_dir
Shutil demonstration complete and cleanup performed.


In [18]:
import json
import os

# 2. Create a Python dictionary with some sample data.
data = {
    'name': 'Alice',
    'age': 30,
    'isStudent': False,
    'courses': [
        {'title': 'Math', 'credits': 3},
        {'title': 'Science', 'credits': 4}
    ]
}

print("Original Python Dictionary:")
print(data)

# 3. Use json.dumps() to convert the Python dictionary to a JSON formatted string and print the result.
json_string = json.dumps(data, indent=4)
print("\nPython Dictionary converted to JSON string:")
print(json_string)

# 4. Use json.loads() to convert the JSON string back to a Python dictionary and print the result.
decoded_data = json.loads(json_string)
print("\nJSON string decoded back to Python Dictionary:")
print(decoded_data)

# 5. Define a filename for a dummy JSON file.
file_name = 'data.json'

# 6. Use open() in write mode ('w') and json.dump() to write the Python dictionary to the dummy JSON file.
with open(file_name, 'w') as f:
    json.dump(data, f, indent=4)
print(f"\nPython Dictionary saved to '{file_name}'.")

# 7. Use open() in read mode ('r') and json.load() to read the JSON data from the dummy JSON file back into a Python variable and print the result.
with open(file_name, 'r') as f:
    loaded_data = json.load(f)
print(f"\nData loaded from '{file_name}':")
print(loaded_data)

# 8. Remove the dummy JSON file using os.remove() to clean up the environment.
os.remove(file_name)
print(f"\nCleaned up: Removed '{file_name}'.")


Original Python Dictionary:
{'name': 'Alice', 'age': 30, 'isStudent': False, 'courses': [{'title': 'Math', 'credits': 3}, {'title': 'Science', 'credits': 4}]}

Python Dictionary converted to JSON string:
{
    "name": "Alice",
    "age": 30,
    "isStudent": false,
    "courses": [
        {
            "title": "Math",
            "credits": 3
        },
        {
            "title": "Science",
            "credits": 4
        }
    ]
}

JSON string decoded back to Python Dictionary:
{'name': 'Alice', 'age': 30, 'isStudent': False, 'courses': [{'title': 'Math', 'credits': 3}, {'title': 'Science', 'credits': 4}]}

Python Dictionary saved to 'data.json'.

Data loaded from 'data.json':
{'name': 'Alice', 'age': 30, 'isStudent': False, 'courses': [{'title': 'Math', 'credits': 3}, {'title': 'Science', 'credits': 4}]}

Cleaned up: Removed 'data.json'.


In [19]:
import csv
import os

# 2. Define a list of lists to represent some sample data
sample_data = [
    ['Name', 'Age', 'City'],
    ['Alice', 30, 'New York'],
    ['Bob', 24, 'Los Angeles'],
    ['Charlie', 35, 'Chicago']
]

# 3. Define a filename for the dummy CSV file
file_name = 'sample.csv'

print("--- Writing data to CSV ---")
# 4. Open the file in write mode ('w') and create a csv.writer object.
# 5. Use the writer.writerow() method to write the header row and then each row of your sample data to the CSV file.
with open(file_name, 'w', newline='') as csvfile:
    csv_writer = csv.writer(csvfile)
    csv_writer.writerows(sample_data)
print(f"Successfully wrote data to '{file_name}'.")

print("\n--- Reading data from CSV ---")
# 7. Open the same CSV file in read mode ('r') and create a csv.reader object.
# 8. Iterate through the csv.reader object and print each row to display the data read from the file.
with open(file_name, 'r', newline='') as csvfile:
    csv_reader = csv.reader(csvfile)
    for row in csv_reader:
        print(row)
print(f"Successfully read data from '{file_name}'.")

# 10. Remove the dummy CSV file using os.remove() to clean up the environment.
os.remove(file_name)
print(f"\nCleaned up: Removed '{file_name}'.")

--- Writing data to CSV ---
Successfully wrote data to 'sample.csv'.

--- Reading data from CSV ---
['Name', 'Age', 'City']
['Alice', '30', 'New York']
['Bob', '24', 'Los Angeles']
['Charlie', '35', 'Chicago']
Successfully read data from 'sample.csv'.

Cleaned up: Removed 'sample.csv'.


In [20]:
import sys

print("--- sys.argv Demonstration ---")
# 1. Print the list of command-line arguments using sys.argv.
# In a Colab notebook, sys.argv usually contains only the path to the script unless explicitly set or run from a shell.
print(f"Current sys.argv: {sys.argv}")
print("Explanation: In a typical Python script run from the command line, sys.argv[0] is the script name, and subsequent elements are arguments passed. In this Colab environment, it usually contains only the path to the executed script file.")

print("\n--- Simulating sys.argv ---")
# Optional: Demonstrate how to simulate command-line arguments for sys.argv
# by temporarily modifying sys.argv for demonstration purposes, then reset it.

# Store the original sys.argv
original_argv = sys.argv

# Simulate command-line arguments
sys.argv = ['script.py', 'arg1', 'arg2', '--option', 'value']
print(f"Simulated sys.argv: {sys.argv}")

# You can now access simulated arguments
print(f"First simulated argument (sys.argv[1]): {sys.argv[1]}")

# Reset sys.argv to its original state to avoid side effects on other parts of the notebook
sys.argv = original_argv
print(f"sys.argv reset to original: {sys.argv}")

print("\n--- sys.platform Demonstration ---")
# 2. Print the operating system platform using sys.platform.
print(f"Operating System Platform: {sys.platform}")
print("Explanation: 'sys.platform' provides a string indicating the platform Python is running on (e.g., 'linux' for Linux, 'win32' for Windows, 'darwin' for macOS).")

--- sys.argv Demonstration ---
Current sys.argv: ['/usr/local/lib/python3.12/dist-packages/colab_kernel_launcher.py', '-f', '/root/.local/share/jupyter/runtime/kernel-27a2b6c2-4fc4-412e-b6d8-e09df37f6d4f.json']
Explanation: In a typical Python script run from the command line, sys.argv[0] is the script name, and subsequent elements are arguments passed. In this Colab environment, it usually contains only the path to the executed script file.

--- Simulating sys.argv ---
Simulated sys.argv: ['script.py', 'arg1', 'arg2', '--option', 'value']
First simulated argument (sys.argv[1]): arg1
sys.argv reset to original: ['/usr/local/lib/python3.12/dist-packages/colab_kernel_launcher.py', '-f', '/root/.local/share/jupyter/runtime/kernel-27a2b6c2-4fc4-412e-b6d8-e09df37f6d4f.json']

--- sys.platform Demonstration ---
Operating System Platform: linux
Explanation: 'sys.platform' provides a string indicating the platform Python is running on (e.g., 'linux' for Linux, 'win32' for Windows, 'darwin' for

In [21]:
import random

print("--- random.randint Demonstration ---")
# 2. Generate a random integer between two specified values (e.g., 1 and 100) using random.randint() and print the result.
random_integer = random.randint(1, 100)
print(f"Random integer between 1 and 100: {random_integer}")

print("\n--- random.choice Demonstration ---")
# 3. Create a list of items (e.g., fruits, numbers).
fruits = ['apple', 'banana', 'cherry', 'date', 'elderberry']
print(f"Original list: {fruits}")

# 4. Select a random item from the list using random.choice() and print the result.
random_fruit = random.choice(fruits)
print(f"Randomly selected fruit: {random_fruit}")

print("\n--- random.shuffle Demonstration ---")
# 5. Create another list of items.
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(f"Original list: {numbers}")

# 6. Shuffle the elements of this list in place using random.shuffle() and print the shuffled list.
random.shuffle(numbers)
print(f"Shuffled list: {numbers}")

--- random.randint Demonstration ---
Random integer between 1 and 100: 62

--- random.choice Demonstration ---
Original list: ['apple', 'banana', 'cherry', 'date', 'elderberry']
Randomly selected fruit: cherry

--- random.shuffle Demonstration ---
Original list: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Shuffled list: [9, 3, 7, 5, 4, 10, 2, 8, 1, 6]


In [22]:
import math

print("--- math.sqrt Demonstration ---")
# 2. Calculate the square root of a number (e.g., 25) using math.sqrt() and print the result.
square_root_result = math.sqrt(25)
print(f"The square root of 25 is: {square_root_result}")

print("\n--- math.pow Demonstration ---")
# 3. Calculate the power of a number (e.g., 2 raised to the power of 3) using math.pow() and print the result.
power_result = math.pow(2, 3)
print(f"2 raised to the power of 3 is: {power_result}")

--- math.sqrt Demonstration ---
The square root of 25 is: 5.0

--- math.pow Demonstration ---
2 raised to the power of 3 is: 8.0
